File: internal/wire/wire_test.go
package wire

import (
	"bytes"
	"testing"
)

func TestEncodeDecodeRoundTrip(t *testing.T) {
	cases := []struct {
		name string
		in   Frame
	}{
		{"output", Frame{Kind: Output, Data: []byte("hello\x1b[0m world")}},
		{"input", Frame{Kind: Input, Data: []byte{0x03}}},
		{"empty output", Frame{Kind: Output, Data: []byte{}}},
		{"resize", Frame{Kind: Resize, Cols: 120, Rows: 40}},
		{"resize zero", Frame{Kind: Resize, Cols: 0, Rows: 0}},
		{"resize max", Frame{Kind: Resize, Cols: 65535, Rows: 65535}},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got, err := Decode(Encode(tc.in))
			if err != nil {
				t.Fatalf("Decode: %v", err)
			}
			if got.Kind != tc.in.Kind {
				t.Errorf("Kind = %d, want %d", got.Kind, tc.in.Kind)
			}
			if got.Cols != tc.in.Cols || got.Rows != tc.in.Rows {
				t.Errorf("dims = %dx%d, want %dx%d", got.Cols, got.Rows, tc.in.Cols, tc.in.Rows)
			}
			if tc.in.Kind != Resize && !bytes.Equal(got.Data, tc.in.Data) {
				t.Errorf("Data = %q, want %q", got.Data, tc.in.Data)
			}
		})
	}
}

func TestDecodeCopiesData(t *testing.T) {
	buf := Encode(Frame{Kind: Output, Data: []byte("abc")})
	got, err := Decode(buf)
	if err != nil {
		t.Fatal(err)
	}
	// Mutating the source buffer must not affect the decoded frame.
	for i := range buf {
		buf[i] = 0
	}
	if !bytes.Equal(got.Data, []byte("abc")) {
		t.Errorf("decoded data aliases source buffer: %q", got.Data)
	}
}

func TestDecodeErrors(t *testing.T) {
	if _, err := Decode(nil); err != ErrShortFrame {
		t.Errorf("empty buffer: got %v, want ErrShortFrame", err)
	}
	if _, err := Decode([]byte{byte(Resize), 0x00}); err != ErrShortFrame {
		t.Errorf("short resize: got %v, want ErrShortFrame", err)
	}
	if _, err := Decode([]byte{0xFF}); err == nil {
		t.Error("unknown kind: expected error, got nil")
	}
}

File: worker/test/wire.test.ts
import { describe, expect, it } from "vitest";
import { decode, encode, Kind, ShortFrameError } from "../src/core/wire";
import { encodeBase32NoPad, newSessionID } from "../src/core/base32";

const te = new TextEncoder();

describe("wire encode", () => {
  it("encodes Output as [kind, ...data]", () => {
    const data = te.encode("hello");
    expect([...encode({ kind: Kind.Output, data })]).toEqual([0, ...data]);
  });

  it("encodes Input as [kind, ...data]", () => {
    const data = te.encode("ls\n");
    expect([...encode({ kind: Kind.Input, data })]).toEqual([1, ...data]);
  });

  it("encodes Resize as kind + two big-endian uint16s", () => {
    // 120 = 0x0078, 40 = 0x0028
    expect([...encode({ kind: Kind.Resize, cols: 120, rows: 40 })]).toEqual([
      2, 0x00, 0x78, 0x00, 0x28,
    ]);
  });

  it("encodes an empty Output frame as a single kind byte", () => {
    expect([...encode({ kind: Kind.Output, data: new Uint8Array(0) })]).toEqual([0]);
  });
});

describe("wire decode", () => {
  it("round-trips Output", () => {
    const data = te.encode("some output bytes \x1b[31m");
    const f = decode(encode({ kind: Kind.Output, data }));
    expect(f.kind).toBe(Kind.Output);
    expect([...f.data!]).toEqual([...data]);
  });

  it("round-trips Resize", () => {
    const f = decode(encode({ kind: Kind.Resize, cols: 200, rows: 50 }));
    expect(f.kind).toBe(Kind.Resize);
    expect(f.cols).toBe(200);
    expect(f.rows).toBe(50);
  });

  it("throws on empty buffer", () => {
    expect(() => decode(new Uint8Array(0))).toThrow(ShortFrameError);
  });

  it("throws on a short Resize payload", () => {
    expect(() => decode(new Uint8Array([2, 0, 1]))).toThrow(ShortFrameError);
  });

  it("throws on an unknown kind", () => {
    expect(() => decode(new Uint8Array([9, 1, 2, 3]))).toThrow();
  });

  it("does not alias the source buffer", () => {
    const src = encode({ kind: Kind.Output, data: te.encode("abc") });
    const f = decode(src);
    src[1] = 0;
    expect(f.data![0]).toBe("a".charCodeAt(0));
  });
});

describe("base32 session IDs", () => {
  it("matches RFC 4648 no-pad vectors (parity with Go StdEncoding)", () => {
    expect(encodeBase32NoPad(te.encode("foobar"))).toBe("MZXW6YTBOI");
    expect(encodeBase32NoPad(new Uint8Array(15))).toBe("A".repeat(24));
    expect(encodeBase32NoPad(new Uint8Array(15).fill(0xff))).toBe("7".repeat(24));
  });

  it("mints 24-character tokens from the alphabet", () => {
    const id = newSessionID();
    expect(id).toHaveLength(24);
    expect(id).toMatch(/^[A-Z2-7]{24}$/);
  });

  it("mints unique IDs", () => {
    const ids = new Set(Array.from({ length: 1000 }, () => newSessionID()));
    expect(ids.size).toBe(1000);
  });
});
